File: airial_api/app/routers/schemas.py
# Endpoints for managing schemas in the API
# Author: Michael B. Lance

# ---------------------------------------------------------------------------------------------------------------------------

from uuid import UUID

from flask import Blueprint, abort, current_app
from flask_login import login_required
from flask_pydantic import validate
from psycopg.errors import DatabaseError

from app.decorators import permission_required
from app.extensions import base
from database.errors import ObjectNotFound, AuthorizationFailure
from database.object_models.project_management.schemas import createSchemaReq

schemaBp = Blueprint("schemas", __name__, url_prefix="/api/v1/schemas")


# ---------------------------------------------------------------------------------------------------------------------------
# GET


@schemaBp.get("/<string:schema_id>/labels")
@login_required
@permission_required("access")
def get_labels(schema_id: str):
    """
    Retrieve labels for a schema
    ---
    responses:
      200:
            description: List of labels.
      404:
            description: No labels / schema found.
      500:
            description: Database error.
    """
    try:
        labels = base.get_schema_labels(UUID(schema_id))
    except ObjectNotFound as e:
        current_app.logger.exception(e)
        abort(404, str(e))
    except AuthorizationFailure as e:
        current_app.logger.exception(e)
        abort(401, str(e))
    except (DatabaseError, Exception) as e:
        current_app.logger.exception(e)
        abort(500)

    return [label.to_dict() for label in labels], 200


# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


@schemaBp.get("<string:schema_id>/models")
@login_required
@permission_required("access")
def get_models(schema_id: str):
    """ """
    try:
        models = base.get_schema_models(UUID(schema_id))
    except ObjectNotFound as e:
        current_app.logger.exception(e)
        abort(404, str(e))
    except AuthorizationFailure as e:
        current_app.logger.exception(e)
        abort(401, str(e))
    except (DatabaseError, Exception) as e:
        current_app.logger.exception(e)
        abort(500)

    return [model.to_dict() for model in models], 200


# ---------------------------------------------------------------------------------------------------------------------------
# POST


@schemaBp.post("")
@login_required
@permission_required("access")
@validate()
def create(body: createSchemaReq):
    """ """
    try:
        schema = base.create_schema(body)

    except (DatabaseError, Exception) as e:
        current_app.logger.exception(e)
        abort(500)

    return schema.to_dict(), 201

File: generator_web_app/src/modules/api/schemas.ts
// API methods for managing schema objects
// Author: Michael B. Lance

// ---------------------------------------------------------------------------------------------------------------------------

import { Label, Model, Schema } from "@/types/generatorobjects.ts";
import type {
  LabelIntf,
  ModelIntf,
  SchemaIntf,
} from "@/types/generatorobjects.ts";
import { ApiError } from "@/modules/api/errors.ts";
import { api_url } from "@/modules/api/apiV1Methods.ts";

// ---------------------------------------------------------------------------------------------------------------------------

export async function getSchemaLabels(schema_id: string): Promise<Label[]> {
  const response = await fetch(`${api_url}/schemas/${schema_id}/labels`, {
    method: "GET",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) throw new ApiError(await response.json());

  const resp = await response.json();
  let labels = [];

  for (const label of resp) labels.push(new Label(label as LabelIntf));

  return labels;
}

// ---------------------------------------------------------------------------------------------------------------------------

export async function getSchemaModels(schema_id: string): Promise<Model[]> {
  const response = await fetch(`${api_url}/schemas/${schema_id}/models`, {
    method: "GET",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) throw new ApiError(await response.json());

  const resp = await response.json();
  let models = [];

  for (const model of resp) models.push(new Model(model as ModelIntf));

  return models;
}

// ---------------------------------------------------------------------------------------------------------------------------
// POST

export interface createSchemaOptions {
  name: string;
  project_id: string;
  model_id: string;
}

export async function createSchema(
  options: createSchemaOptions,
): Promise<Schema> {
  const response = await fetch(`${api_url}/schemas`, {
    method: "POST",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(options),
  });

  if (!response.ok) throw new ApiError(await response.json());

  return new Schema((await response.json()) as SchemaIntf);
}

// ---------------------------------------------------------------------------------------------------------------------------
// DELETE

export async function deleteSchema(schema_id: string): Promise<boolean> {
  const response = await fetch(`${api_url}/schemas/${schema_id}`, {
    method: "DELETE",
    credentials: "include",
    headers: {
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) throw new ApiError(await response.json());

  return true;
}
